File: sdk/python/sen/transport/ndjson.py
"""NDJSON transport — spawns ``sen agent --output-format stream-json``
and communicates over stdin/stdout using newline-delimited JSON."""

from __future__ import annotations

import asyncio
import json
import logging
from typing import Any, Callable, Optional

from sen.ndjson_models import (
    ControlRequest,
    ControlResponse,
    StdinMessage,
    StdoutMessage,
    parse_stdout_message,
)

log = logging.getLogger(__name__)


class NdjsonTransport:
    """Async NDJSON transport over a ``sen`` subprocess.

    This is a standalone transport that speaks the streaming NDJSON protocol
    (``--output-format stream-json``), **not** JSON-RPC.  It does not inherit
    from :class:`~sen.transport.base.Transport`.

    Usage::

        async with NdjsonTransport(binary="sen") as t:
            await t.send(UserMessage(content="Hello"))
            msg = await t.recv()
    """

    def __init__(
        self,
        binary: str = "sen",
        args: Optional[list[str]] = None,
        cwd: Optional[str] = None,
        on_permission: Optional[Callable[..., Any]] = None,
        env: Optional[dict[str, str]] = None,
    ) -> None:
        self._binary = binary
        self._extra_args: list[str] = args or []
        self._cwd = cwd
        self._on_permission = on_permission
        self._env = env

        self._proc: Optional[asyncio.subprocess.Process] = None
        self._queue: asyncio.Queue[StdoutMessage] = asyncio.Queue()
        self._read_task: Optional[asyncio.Task[None]] = None
        self._closed = False

    # ── lifecycle ─────────────────────────────────────────────────────

    async def start(self) -> None:
        """Spawn the subprocess and begin reading stdout."""
        cmd = [
            self._binary,
            "agent",
            "--output-format",
            "stream-json",
            *self._extra_args,
        ]
        self._proc = await asyncio.create_subprocess_exec(
            *cmd,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            cwd=self._cwd,
            env=self._env,
        )
        self._read_task = asyncio.create_task(self._read_loop())

    async def close(self) -> None:
        """Terminate the subprocess and cancel the reader."""
        if self._closed:
            return
        self._closed = True

        if self._read_task and not self._read_task.done():
            self._read_task.cancel()
            try:
                await self._read_task
            except asyncio.CancelledError:
                pass

        if self._proc:
            try:
                self._proc.terminate()
                await asyncio.wait_for(self._proc.wait(), timeout=5)
            except (ProcessLookupError, asyncio.TimeoutError):
                self._proc.kill()
            self._proc = None

    async def __aenter__(self) -> NdjsonTransport:
        await self.start()
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.close()

    # ── send / recv ───────────────────────────────────────────────────

    async def send(self, msg: StdinMessage) -> None:
        """Write a :class:`StdinMessage` as a single NDJSON line to stdin."""
        if self._proc is None or self._proc.stdin is None:
            raise RuntimeError("Transport not started")

        raw = json.dumps(msg.to_dict(), ensure_ascii=False)
        raw = raw.replace("\u2028", "\\u2028").replace("\u2029", "\\u2029")
        self._proc.stdin.write((raw + "\n").encode("utf-8"))
        await self._proc.stdin.drain()

    async def recv(self) -> StdoutMessage:
        """Return the next parsed :class:`StdoutMessage` from the queue.

        Blocks until a message is available.  ``control_request`` messages
        are handled internally by the read loop and never appear here unless
        no ``on_permission`` callback was provided.
        """
        return await self._queue.get()

    # ── internal read loop ────────────────────────────────────────────

    async def _read_loop(self) -> None:
        assert self._proc is not None and self._proc.stdout is not None
        try:
            while True:
                line = await self._proc.stdout.readline()
                if not line:
                    break
                line_str = line.decode("utf-8", errors="replace").strip()
                if not line_str:
                    continue
                try:
                    data = json.loads(line_str)
                except json.JSONDecodeError:
                    log.warning("Ignoring non-JSON line: %s", line_str[:120])
                    continue

                msg = parse_stdout_message(data)

                if isinstance(msg, ControlRequest) and self._on_permission is not None:
                    await self._handle_control_request(msg)
                else:
                    await self._queue.put(msg)
        except asyncio.CancelledError:
            return
        except Exception:
            log.exception("NdjsonTransport read loop crashed")

    async def _handle_control_request(self, req: ControlRequest) -> None:
        """Invoke the ``on_permission`` callback and send the response."""
        assert self._on_permission is not None
        try:
            result = self._on_permission(req)
            if asyncio.iscoroutine(result):
                result = await result

            if isinstance(result, ControlResponse):
                response = result
            elif result is True or result is None:
                response = ControlResponse(
                    request_id=req.request_id, decision="allow"
                )
            elif result is False:
                response = ControlResponse(
                    request_id=req.request_id,
                    decision="deny",
                    reason="denied by SDK callback",
                )
            else:
                response = ControlResponse(
                    request_id=req.request_id, decision="allow"
                )
        except Exception as exc:
            log.error("on_permission callback raised: %s — denying by default", exc)
            response = ControlResponse(
                request_id=req.request_id, decision="deny"
            )

        await self.send(response)

File: sdk/typescript/src/ndjson.ts
// SPDX-License-Identifier: MIT
// NDJSON transport for headless / SDK mode.
// Speaks the streaming NDJSON protocol (--output-format stream-json), NOT JSON-RPC.

import { spawn, type ChildProcess } from "node:child_process";
import { createInterface } from "node:readline";
import type {
  StdinMessage,
  StdoutMessage,
  ControlRequestCanUseTool,
  ControlRequestSessionStateChanged,
  ControlRequestMcpSetServers,
  ControlResponseAllow,
  ControlResponseDeny,
} from "./types.js";

export type PermissionDecision =
  | ControlResponseAllow
  | ControlResponseDeny
  | boolean;

export interface NdjsonOptions {
  binary?: string;
  args?: string[];
  cwd?: string;
  env?: Record<string, string>;
  onPermission?: (
    req: ControlRequestCanUseTool | ControlRequestSessionStateChanged | ControlRequestMcpSetServers,
  ) => PermissionDecision | Promise<PermissionDecision>;
}

export class NdjsonTransport {
  private proc: ChildProcess | null = null;
  private queue: StdoutMessage[] = [];
  private waiters: Array<(msg: StdoutMessage) => void> = [];
  private closed = false;

  private readonly binary: string;
  private readonly args: string[];
  private readonly cwd?: string;
  private readonly env?: Record<string, string>;
  private readonly onPermission?: NdjsonOptions["onPermission"];

  constructor(options?: NdjsonOptions) {
    this.binary = options?.binary ?? "sen";
    this.args = options?.args ?? [];
    this.cwd = options?.cwd;
    this.env = options?.env;
    this.onPermission = options?.onPermission;
  }

  async start(): Promise<void> {
    const cmd = ["agent", "--output-format", "stream-json", ...this.args];
    this.proc = spawn(this.binary, cmd, {
      // Do NOT merge with process.env when env is empty — that causes the
      // subprocess to lose PATH and other essential variables on Windows.
      stdio: ["pipe", "pipe", "pipe"],
      cwd: this.cwd,
      env: this.env ? { ...process.env, ...this.env } : undefined,
    });

    // Drain stderr to prevent the subprocess from blocking on EPIPE.
    this.proc.stderr?.on("data", (chunk: Buffer) => {
      process.stderr.write(chunk);
    });

    const rl = createInterface({ input: this.proc.stdout! });
    rl.on("line", (line) => {
      const trimmed = line.trim();
      if (!trimmed) return;

      let data: Record<string, unknown>;
      try {
        data = JSON.parse(trimmed);
      } catch {
        return;
      }

      const msg = data as unknown as StdoutMessage;
      if (msg.type === "control_request" && this.onPermission) {
        void this.handleControlRequest(msg as Parameters<typeof this.handleControlRequest>[0]);
      } else {
        this.enqueue(msg);
      }
    });

    // Immediately check for startup failures — spawn() calls back with an
    // error event synchronously if the binary does not exist.
    this.proc.on("error", (err) => {
      process.stderr.write(`[NdjsonTransport] subprocess error: ${err.message}\n`);
    });

    this.proc.on("close", () => {
      this.closed = true;
    });
  }

  async send(msg: StdinMessage): Promise<void> {
    if (!this.proc?.stdin?.writable) {
      throw new Error("Transport not started");
    }
    const raw = JSON.stringify(msg)
      .replace(/\u2028/g, "\\u2028")
      .replace(/\u2029/g, "\\u2029");
    this.proc.stdin.write(raw + "\n");
  }

  async *recv(): AsyncGenerator<StdoutMessage> {
    while (!this.closed || this.queue.length > 0) {
      if (this.queue.length > 0) {
        yield this.queue.shift()!;
      } else {
        const msg = await new Promise<StdoutMessage>((resolve) => {
          this.waiters.push(resolve);
        });
        yield msg;
      }
    }
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;

    if (this.proc) {
      this.proc.kill("SIGTERM");
      await new Promise<void>((resolve) => {
        const timer = setTimeout(() => {
          this.proc?.kill("SIGKILL");
          resolve();
        }, 5_000);
        this.proc!.on("close", () => {
          clearTimeout(timer);
          resolve();
        });
      });
      this.proc = null;
    }

    for (const waiter of this.waiters) {
      waiter({ type: "system", content: "transport closed" });
    }
    this.waiters = [];
  }

  private enqueue(msg: StdoutMessage): void {
    if (this.waiters.length > 0) {
      this.waiters.shift()!(msg);
    } else {
      this.queue.push(msg);
    }
  }

  private buildControlResponse(
    req: ControlRequestCanUseTool | ControlRequestSessionStateChanged | ControlRequestMcpSetServers,
    result: PermissionDecision,
  ): StdinMessage {
    // boolean → explicit allow/deny shape
    if (typeof result === "boolean") {
      return result
        ? ({ type: "control_response", request_id: req.request_id, decision: "allow" } as const)
        : ({
            type: "control_response",
            request_id: req.request_id,
            decision: "deny",
            reason: "denied by SDK callback",
          } as const);
    }
    // Already a typed shape
    return result as StdinMessage;
  }

  private async handleControlRequest(
    req: ControlRequestCanUseTool | ControlRequestSessionStateChanged | ControlRequestMcpSetServers,
  ): Promise<void> {
    if (!this.onPermission) return;

    let response: StdinMessage;
    try {
      const result = await Promise.resolve(this.onPermission(req));
      response = this.buildControlResponse(req, result);
    } catch {
      // Deny on exception — always safer than implicitly allowing.
      response = {
        type: "control_response",
        request_id: req.request_id,
        decision: "deny",
        reason: "permission callback raised an exception",
      } as const;
    }

    await this.send(response);
  }
}
